ipatests/test_integration/test_hbac_functional.py

```diff
@@ -0,0 +1,673 @@
+        ])
+        self.master.run_command([
+            "ipa", "hbacrule-add-user", "admin_allow_all", 
+            "--groups=admins"
+        ])
+        self.master.run_command(["ipa", "hbacrule-disable", "allow_all"])
+
+    def cleanup_rule_and_groups(self, rule_name, *group_names):
+        """Helper to clean up rules and groups"""
+        tasks.kinit_admin(self.master)
+        self.master.run_command([
+            "ipa", "hbacrule-del", rule_name
+        ], raiseonerr=False)
+        for group_name in group_names:
+            if group_name.startswith('hostgrp') or 'hostgroup' in group_name:
+                self.master.run_command([
+                    "ipa", "hostgroup-del", group_name
+                ], raiseonerr=False)
+            elif group_name.startswith('svcgrp') or 'svcgroup' in group_name:
+                self.master.run_command([
+                    "ipa", "hbacsvcgroup-del", group_name
+                ], raiseonerr=False)
+            else:
+                self.master.run_command([
+                    "ipa", "group-del", group_name
+                ], raiseonerr=False)
+
+    def test_hbacsvc_master_001(self):
+        """
+        hbacsvc_master_001: user access to client, on master, add rule
+        """
+        # Use existing test users
+        user_allowed = self.TEST_USERS[0]  # testuser1
+        user_denied = self.TEST_USERS[1]   # testuser2
+        
+        # SSH auth success tests as in bash
+        result = self.master.run_command([
+            'sshpass', '-p', self.USER_PASSWORD, 'ssh', '-o', 
+            'StrictHostKeyChecking=no', '-l', user_allowed, 
+            self.master.hostname, 'whoami'
+        ], raiseonerr=False)
+        
+        # Setup HBAC rule on master
+        self.setup_admin_rule_and_disable_allow_all()
+        
+        self.master.run_command(["ipa", "hbacrule-add", "rule1"])
+        self.master.run_command([
+            "ipa", "hbacrule-add-user", "rule1", f"--users={user_allowed}"
+        ])
+        self.master.run_command([
+            "ipa", "hbacrule-add-host", "rule1", 
+            f"--hosts={self.clients[0].hostname}"
+        ])
+        self.master.run_command([
+            "ipa", "hbacrule-add-service", "rule1", "--hbacsvcs=sshd"
+        ])
+        self.master.run_command(["ipa", "hbacrule-show", "rule1", "--all"])
+        
+        # Test using helper
+        self.assert_hbac_access(user_allowed, self.clients[0].hostname, 
+                               "sshd", should_allow=True, expected_rule="rule1")
+        self.assert_hbac_access(user_denied, self.clients[0].hostname, 
+                               "sshd", should_allow=False)
+        self.assert_hbac_access(user_allowed, self.clients[1].hostname, 
+                               "sshd", should_allow=False)
+
+    def test_hbacsvc_master_001_cleanup(self):
+        """Cleanup for hbacsvc_master_001"""
+        self.cleanup_rule_and_groups("rule1", "admin_allow_all")
```

Comment:
Rather than as a separate, fake test, can the cleanup be done as a fixture instead, so that after each test is executed cleanup happens? I guess at least move it to the uninstall classmethod override?